### DISCUSSION TITLE
Renovate not automerging PR, even when `Automerge: Enabled` in PR

### DISCUSSION THREAD

**Discussion by therealvio** [renovatebot/renovate#20841] [Request Help] [answered]
### I'm using the Mend Renovate hosted app on github.com.

- [X] I'm using the Mend Renovate hosted app.

### Was this something which used to work for you, and then stopped?

It used to work, and then stopped

### Describe your problem.

Hi team,

Around 7 days ago, `renovate` has stopped automerging my PRs. Although there was a configuration change a few days prior, PRs were still automerging just fine until then.

For example, in [this PR](https://github.com/therealvio-org/ralphbot/pull/328), it indicates that the PR is set up for automerging, whereas in the logs for the most-recent job at the time, it states that the PR's associated branch is *not* set up for automerging (snippet attached).

I am a little confused what may have happened that caused this, as the configuration was previously working, unless a certain property has been deprecated and that's what is causing a stir. The configuration for my project [is here](https://github.com/therealvio-org/ralphbot/blob/main/.github/renovate.json).

Any guidance would be appreciated.

Thanks again for this great project!

### Relevant debug logs.

<details><summary>Logs</summary>

```
DEBUG: syncBranchState()(branch="renovate/constructs-10.x")
DEBUG: syncBranchState(): update baseBranchSha(branch="renovate/constructs-10.x")
DEBUG: syncBranchState(): update branchSha(branch="renovate/constructs-10.x")
DEBUG: getBranchPr(renovate/constructs-10.x)(branch="renovate/constructs-10.x")
DEBUG: findPr(renovate/constructs-10.x, undefined, open)(branch="renovate/constructs-10.x")
DEBUG: Found PR #328(branch="renovate/constructs-10.x")
DEBUG: branchExists=true(branch="renovate/constructs-10.x")
DEBUG: dependencyDashboardCheck=undefined(branch="renovate/constructs-10.x")
DEBUG: PR rebase requested=false(branch="renovate/constructs-10.x")
DEBUG: Checking if PR has been edited(branch="renovate/constructs-10.x")
DEBUG: branch.isModified(): using git to calculate(branch="renovate/constructs-10.x")
DEBUG: branch.isModified() = false(branch="renovate/constructs-10.x")
DEBUG: Found existing branch PR(branch="renovate/constructs-10.x")
DEBUG: Checking schedule(after 8pm every weekday,before 8am every weekday,every weekend, Australia/Melbourne)(branch="renovate/constructs-10.x")
DEBUG: Found timezone: Australia/Melbourne(branch="renovate/constructs-10.x")
DEBUG: Adjusting now for timezone(branch="renovate/constructs-10.x")
DEBUG: Checking 3 schedule(s)(branch="renovate/constructs-10.x")
DEBUG: Checking schedule "after 8pm every weekday"(branch="renovate/constructs-10.x")
{
  "parsedSchedule": {
    "schedules": [
      {
        "t_a": [
          72000
        ],
        "d": [
          2,
          3,
          4,
          5,
          6
        ]
      }
    ],
    "exceptions": [],
    "error": -1
  }
}
DEBUG: Matches schedule after 8pm every weekday(branch="renovate/constructs-10.x")
DEBUG: Branch already exists(branch="renovate/constructs-10.x")
DEBUG: getBranchPr(renovate/constructs-10.x)(branch="renovate/constructs-10.x")
DEBUG: findPr(renovate/constructs-10.x, undefined, open)(branch="renovate/constructs-10.x")
DEBUG: Found PR #328(branch="renovate/constructs-10.x")
DEBUG: branch.isBehindBase(): using git to calculate(branch="renovate/constructs-10.x")
DEBUG: branch.isBehindBase(): false(branch="renovate/constructs-10.x")
{
  "currentBranch": "main",
  "currentBranchSha": "ea935399dced561e5686ee12dc0fbad9f9fcea8c"
}
DEBUG: Branch is up-to-date(branch="renovate/constructs-10.x")
DEBUG: isBranchConflicted(main, renovate/constructs-10.x)(branch="renovate/constructs-10.x")
DEBUG: branch.isConflicted(): using git to calculate(branch="renovate/constructs-10.x")
DEBUG: branch.isConflicted(): false(branch="renovate/constructs-10.x")
DEBUG: Branch does not need rebasing(branch="renovate/constructs-10.x")
DEBUG: Using reuseExistingBranch: true(branch="renovate/constructs-10.x")
DEBUG: Setting current branch to main(branch="renovate/constructs-10.x")
DEBUG: latest commit(branch="renovate/constructs-10.x")
{
  "branchName": "main",
  "latestCommitDate": "2023-03-10T21:04:12+11:00"
}
DEBUG: manager.getUpdatedPackageFiles() reuseExistingBranch=true(branch="renovate/constructs-10.x")
DEBUG: npm.updateDependency(): dependencies.constructs = 10.1.273(branch="renovate/constructs-10.x")
DEBUG: No package files need updating(branch="renovate/constructs-10.x")
DEBUG: Getting updated lock files(branch="renovate/constructs-10.x")
DEBUG: Writing package.json files(branch="renovate/constructs-10.x")
{
  "packageFiles": [
    "ops/package.json"
  ]
}
DEBUG: Writing any updated package files(branch="renovate/constructs-10.x")
DEBUG: No updated lock files in branch(branch="renovate/constructs-10.x")
DEBUG: No files to commit(branch="renovate/constructs-10.x")
DEBUG: Setting current branch to main(branch="renovate/constructs-10.x")
DEBUG: latest commit(branch="renovate/constructs-10.x")
{
  "branchName": "main",
  "latestCommitDate": "2023-03-10T21:04:12+11:00"
}
DEBUG: Checking if we can automerge branch(branch="renovate/constructs-10.x")
DEBUG: mergeStatus=no automerge(branch="renovate/constructs-10.x")
DEBUG: Ensuring PR(branch="renovate/constructs-10.x")
DEBUG: There are 0 errors and 0 warnings(branch="renovate/constructs-10.x")
DEBUG: getBranchPr(renovate/constructs-10.x)(branch="renovate/constructs-10.x")
DEBUG: findPr(renovate/constructs-10.x, undefined, open)(branch="renovate/constructs-10.x")
DEBUG: Found PR #328(branch="renovate/constructs-10.x")
DEBUG: getPrCache()(branch="renovate/constructs-10.x")
DEBUG: Found existing PR(branch="renovate/constructs-10.x")
DEBUG: PR fingerprints mismatch, processing PR(branch="renovate/constructs-10.x")
DEBUG: Fetching changelog: https://github.com/aws/constructs (10.1.264 -> 10.1.273)(branch="renovate/constructs-10.x")
DEBUG: Processing existing PR(branch="renovate/constructs-10.x")
DEBUG: resolveBranchStatus(branchName=renovate/constructs-10.x, ignoreTests=false)(branch="renovate/constructs-10.x")
DEBUG: getBranchStatus(renovate/constructs-10.x)(branch="renovate/constructs-10.x")
DEBUG: branch status check result(branch="renovate/constructs-10.x")
{
  "state": "pending",
  "statuses": []
}
DEBUG: check runs result(branch="renovate/constructs-10.x")
{
  "checkRuns": [
    {
      "name": "ralphbot-cdk-stack-test (test)",
      "status": "in_progress",
      "conclusion": null
    },
    {
      "name": "ralphbot-cdk-stack-test (tsc)",
      "status": "completed",
      "conclusion": "success"
    },
    {
      "name": "ralphbot-cdk-stack-test (eslint)",
      "status": "in_progress",
      "conclusion": null
    },
    {
      "name": "ralphbot-cdk-stack-test (prettier)",
      "status": "completed",
      "conclusion": "success"
    }
  ]
}
DEBUG: Branch status yellow(branch="renovate/constructs-10.x")
DEBUG: setPrCache()(branch="renovate/constructs-10.x")
DEBUG: Pull Request #328 does not need updating(branch="renovate/constructs-10.x")
DEBUG: PR is configured for automerge(branch="renovate/constructs-10.x")
DEBUG: checking auto-merge(branch="renovate/constructs-10.x")
DEBUG: Checking schedule(after 8pm every weekday,before 8am every weekday,every weekend, Australia/Melbourne)(branch="renovate/constructs-10.x")
DEBUG: Found timezone: Australia/Melbourne(branch="renovate/constructs-10.x")
DEBUG: Adjusting now for timezone(branch="renovate/constructs-10.x")
DEBUG: Checking 3 schedule(s)(branch="renovate/constructs-10.x")
DEBUG: Checking schedule "after 8pm every weekday"(branch="renovate/constructs-10.x")
{
  "parsedSchedule": {
    "schedules": [
      {
        "t_a": [
          72000
        ],
        "d": [
          2,
          3,
          4,
          5,
          6
        ]
      }
    ],
    "exceptions": [],
    "error": -1
  }
}
DEBUG: Matches schedule after 8pm every weekday(branch="renovate/constructs-10.x")
DEBUG: resolveBranchStatus(branchName=renovate/constructs-10.x, ignoreTests=false)(branch="renovate/constructs-10.x")
DEBUG: getBranchStatus(renovate/constructs-10.x)(branch="renovate/constructs-10.x")
DEBUG: branch status check result(branch="renovate/constructs-10.x")
{
  "state": "pending",
  "statuses": []
}
DEBUG: check runs result(branch="renovate/constructs-10.x")
{
  "checkRuns": [
    {
      "name": "ralphbot-cdk-stack-test (test)",
      "status": "in_progress",
      "conclusion": null
    },
    {
      "name": "ralphbot-cdk-stack-test (tsc)",
      "status": "completed",
      "conclusion": "success"
    },
    {
      "name": "ralphbot-cdk-stack-test (eslint)",
      "status": "in_progress",
      "conclusion": null
    },
    {
      "name": "ralphbot-cdk-stack-test (prettier)",
      "status": "completed",
      "conclusion": "success"
    }
  ]
}
DEBUG: Branch status yellow(branch="renovate/constructs-10.x")
DEBUG: PR is not ready for merge (branch status is yellow)(branch="renovate/constructs-10.x")
```

</details>

**Comment by rarkins** [comment, marked as answer]
It actually says:

```
DEBUG: PR is configured for automerge(branch="renovate/constructs-10.x")
```

**Reply by therealvio** [reply]
It's most likely I misunderstand, though what does this logline mean then?
`mergeStatus=no automerge(branch="renovate/constructs-10.x")`

**Reply by rarkins** [reply]
That refers to branch automerge, not PR automerge

**Reply by therealvio** [reply]
So with this in mind, the next thing I can think of that is causing a problem is the "review required" branch protection for PRs.

Though I understood that `renovate` should still be merging since I have [added renovate to the bypass required PRs list](https://github.blog/changelog/2021-11-19-allow-bypassing-required-pull-requests/), since I have had this configuration for as long as I had renovate setup. So I am not quite sure why this could be a problem now.

For now, I have set up `renovate-approve` for the repo, though it hasn't approved the PRs that are marked as "Automerge: Enabled". I imagine it is because it doesn't work with old PRs. So in this instance, I am going to merge these PRs myself and see what happens down the road. 🕵️‍♂️

Edit: Grammar, and reworded reference to the "bypass required PRs" repo configuration.

**Reply by rarkins** [reply]
https://docs.renovatebot.com/key-concepts/automerge/

**Reply by therealvio** [reply]
Thanks for the documents, though I am afraid I have read through it while troubleshooting this issue (hence why I have included `renovate-approve` as a means to see if this fixes my problem).

In addition:

- I have globally enabled `platformAutomerge`, enabled `automerge` on what dependencies I would like to automerge;
- As depicted in the linked PR, `automerge` is enabled, so I imagine my configuration is correct [based on this section](https://docs.renovatebot.com/key-concepts/automerge/#automerge-not-enabled-correctly-in-config)
- I have already permitted `renovate` to bypass required pull requests here (pictured)
<img width="757" alt="image" src="https://user-images.githubusercontent.com/41095688/224532391-d990c508-f682-46f9-90ea-76d14f1e4fe6.png">

As I said, things were working a about a week ago, and I am unsure what has changed to cause `renovate` to not automerge PRs anymore. Is there something in particular you believe that I should be paying extra attention to @rarkins? 

Edit: Taking the edge out of my reply.

**Reply by therealvio** [reply]
Oh, and also for others checking out this discussion: Some PRs have been up for as long as 10 days at the time of the original post. No commits were being made to the `main` branch of the repo, so I doubt it had something to do with `renovate` struggling to find an opportunity to merge its PRs because it had to "recalculate" as described in the first [section here](https://docs.renovatebot.com/key-concepts/automerge/#introduction). Otherwise, I have been giving `renovate` room for it to do its thing.

**Reply by rarkins** [reply]
1. Wait or trigger a new automerging PR somehow (e.g. rename and close an existing one)
2. Observe logs at creation (should see platform automerge enabled)
3. Wait for tests to pass

Platform automerge is the best approach because it doesn't require Renovate to be running again after step 3 above - GitHub itself should just merge it.

Platform automerge won't apply to existing PRs - only at time of PR creation.

**Reply by therealvio** [reply]
tl;dr: I managed to get `renovate` to perform its `automerge` after paring down some configurations both on `renovate` and Github. I may have had `automerge` working by coincidence for a while (maybe I ran into some bug that was fixed? I would need to look more into the changes to `renovate` when my problem started showing up), which would have contributed to my confusion

I gave this a go, and both my config and the logs show that "automerge" is enabled (snippet at end of reply). Though 11-ish hours later and [still no automerge occurred](https://github.com/therealvio-org/ralphbot/pull/342) (same with [lockfile maintenance](https://github.com/therealvio-org/ralphbot/pull/341)). `platformAutomerge` was already enabled prior too. I got a feeling that something is off with `renovate`'s automerge, whereas the `platformAutomerge` is working fine.

- I have ran [into this past discussion](https://github.com/renovatebot/renovate/discussions/13273) and compared apples-to-apples where I may be going wrong and from a high-level point of view, my config for `renovate` and Github tracks.
- For the sake of testing, I have removed required reviewers, and disabled `platformAutomerge`, and set the `automergeSchedule` to any time. I triggered [another PR](https://github.com/therealvio-org/ralphbot/pull/343). As of writing this, Renovate-automerge (where `renovate` is doing the work) **successfully ran**, whereas [Github's platformAutomerge](https://docs.renovatebot.com/configuration-options/#platformautomerge) is working as expected.
  - This would mean, that the old behaviour I somehow had working no longer works. Without checking the code for `renovate` itself, maybe I ran into a bug, and it got fixed 🤷 . I will take a closer look later (this is now my own problem). 
 

To make sure I am not misunderstanding going forward: Based on what [this section describes in your documents](https://github.com/therealvio-org/ralphbot/pull/343), required reviews *need* an approval so renovate cannot run its own `automerge` even if `renovate` is supposedly allowed to bypass reviews. So, it has to only rely on `platformAutomerge` to get the job done. Is that right?

The point is: I got something working again, and I think going forward I can still enable "required reviews", just that now I *do* require the use of `renovate-approve` to help with automating my dependency management strategy, rather than having something that worked by coincidence.

Thanks for your time @rarkins, I really appreciate it, and to everyone else who works on this project! ❤️ 

<details><summary>Log Snippet</summary>
```
DEBUG: massaged config
{
  "config": {
    "$schema": "https://docs.renovatebot.com/renovate-schema.json",
    "extends": [
      "github>whitesource/merge-confidence:beta",
      "config:base",
      ":dependencyDashboard"
    ],
    "timezone": "Australia/Melbourne",
    "schedule": [
      "after 8pm every weekday",
      "before 8am every weekday",
      "every weekend"
    ],
    "automergeSchedule": [
      "after 8pm every weekday",
      "before 8am every weekday",
      "every weekend"
    ],
    "postUpdateOptions": [
      "gomodTidy",
      "gomodUpdateImportPaths"
    ],
    "lockFileMaintenance": {
      "enabled": true,
      "schedule": [
        "after 7am on the first day of the month"
      ],
      "automerge": true,
      "automergeType": "pr",
      "platformAutomerge": true,
      "automergeStrategy": "squash"
    },
    "platformAutomerge": true,
    "automergeStrategy": "squash",
    "automergeType": "pr",
    "packageRules": [
      {
        "automerge": true,
        "matchManagers": [
          "npm"
        ],
        "excludeDepPatterns": [
          "node"
        ],
        "matchUpdateTypes": [
          "minor",
          "patch",
          "rollback",
          "bump"
        ]
      },
      {
        "matchManagers": [
          "github-actions"
        ],
        "automerge": true
      },
      {
        "matchManagers": [
          "gomod"
        ],
        "automerge": false
      },
      {
        "groupName": "aws-cdk",
        "automerge": true,
        "allowedVersions": "^2.0.0",
        "matchPackageNames": [
          "aws-cdk",
          "aws-cdk-lib"
        ]
      },
      {
        "groupName": "golang",
        "automerge": false,
        "matchDatasources": [
          "golang-version"
        ],
        "matchManagers": [
          "asdf",
          "dockerfile",
          "gomod"
        ],
        "rangeStrategy": "bump"
      },
      {
        "groupName": "linting",
        "automerge": true,
        "matchPackageNames": [
          "@typescript-eslint/eslint-plugin",
          "@typescript-eslint/parser",
          "eslint",
          "eslint-plugin-import",
          "prettier"
        ]
      },
      {
        "groupName": "jest",
        "automerge": true,
        "matchPackageNames": [
          "jest",
          "ts-jest",
          "@types/jest"
        ]
      },
      {
        "groupName": "nodejs",
        "automerge": false,
        "allowedVersions": "^18.0.0",
        "matchManagers": [
          "npm"
        ],
        "matchPackageNames": [
          "node",
          "@types/node"
        ]
      },
      {
        "groupName": "nodejs",
        "automerge": false,
        "allowedVersions": "^18.0.0",
        "matchDatasources": [
          "node"
        ],
        "matchManagers": [
          "asdf"
        ],
        "matchPackageNames": [
          "node"
        ]
      },
      {
        "groupName": "nodejs",
        "automerge": false,
        "matchManagers": [
          "dockerfile"
        ],
        "matchPackageNames": [
          "node"
        ],
        "matchUpdateTypes": [
          "minor"
        ]
      }
    ]
  }
}
DEBUG: migrated config
{
  "config": {
    "$schema": "https://docs.renovatebot.com/renovate-schema.json",
    "extends": [
      "github>whitesource/merge-confidence:beta",
      "config:base",
      ":dependencyDashboard"
    ],
    "timezone": "Australia/Melbourne",
    "schedule": [
      "after 8pm every weekday",
      "before 8am every weekday",
      "every weekend"
    ],
    "automergeSchedule": [
      "after 8pm every weekday",
      "before 8am every weekday",
      "every weekend"
    ],
    "postUpdateOptions": [
      "gomodTidy",
      "gomodUpdateImportPaths"
    ],
    "lockFileMaintenance": {
      "enabled": true,
      "schedule": [
        "after 7am on the first day of the month"
      ],
      "automerge": true,
      "automergeType": "pr",
      "platformAutomerge": true,
      "automergeStrategy": "squash"
    },
    "platformAutomerge": true,
    "automergeStrategy": "squash",
    "automergeType": "pr",
    "packageRules": [
      {
        "automerge": true,
        "matchManagers": [
          "npm"
        ],
        "excludeDepPatterns": [
          "node"
        ],
        "matchUpdateTypes": [
          "minor",
          "patch",
          "rollback",
          "bump"
        ]
      },
      {
        "matchManagers": [
          "github-actions"
        ],
        "automerge": true
      },
      {
        "matchManagers": [
          "gomod"
        ],
        "automerge": false
      },
      {
        "groupName": "aws-cdk",
        "automerge": true,
        "allowedVersions": "^2.0.0",
        "matchPackageNames": [
          "aws-cdk",
          "aws-cdk-lib"
        ]
      },
      {
        "groupName": "golang",
        "automerge": false,
        "matchDatasources": [
          "golang-version"
        ],
        "matchManagers": [
          "asdf",
          "dockerfile",
          "gomod"
        ],
        "rangeStrategy": "bump"
      },
      {
        "groupName": "linting",
        "automerge": true,
        "matchPackageNames": [
          "@typescript-eslint/eslint-plugin",
          "@typescript-eslint/parser",
          "eslint",
          "eslint-plugin-import",
          "prettier"
        ]
      },
      {
        "groupName": "jest",
        "automerge": true,
        "matchPackageNames": [
          "jest",
          "ts-jest",
          "@types/jest"
        ]
      },
      {
        "groupName": "nodejs",
        "automerge": false,
        "allowedVersions": "^18.0.0",
        "matchManagers": [
          "npm"
        ],
        "matchPackageNames": [
          "node",
          "@types/node"
        ]
      },
      {
        "groupName": "nodejs",
        "automerge": false,
        "allowedVersions": "^18.0.0",
        "matchDatasources": [
          "node"
        ],
        "matchManagers": [
          "asdf"
        ],
        "matchPackageNames": [
          "node"
        ]
      },
      {
        "groupName": "nodejs",
        "automerge": false,
        "matchManagers": [
          "dockerfile"
        ],
        "matchPackageNames": [
          "node"
        ],
        "matchUpdateTypes": [
          "minor"
        ]
      }
    ]
  }
}
```

**Reply by wyardley** [reply]
Having essentially the same issue. I will test `platformAutomerge` as well, but curious about whether either `platformAutomerge` _or_ renovate will work with the "Allow specified actors" box checked, but also "Require review from Code Owners" is set (assuming that the "enforce checks for all users" box is also not ticked).

With non-platform native automerge, I see
```
DEBUG: GitHub blocking PR merge -- will keep trying(branch="renovate/pylint-2.x")
{
  "response": {
    "message": "Waiting on code owner review from org/team.",
    "documentation_url": "https://docs.github.com/articles/about-protected-branches"
  }
}
```
But I think Renovate should be able to bypass this check because of the other setting?

I can't use the branch merge option at least on many projects, among other things, because I've configured our CI system to only run on branches with open PRs.

**Reply by rarkins** [reply]
I think you see that message only when GitHub has rejected our attempt to merge. i.e. no, Renovate apparently is not allowed to bypass it

**Reply by viceice** [reply]
- #21167

**Reply by therealvio** [reply]
~I noticed [this](https://github.blog/changelog/2022-05-17-consistently-allow-github-apps-as-exceptions-to-branch-protection-rules/) was recently (less than 2 weeks ago) announced. I am curious if it impactd `renovate` and makes my old configuration scheme work again.~

~I will check it out later hopefully around 6pm AEST (barring life-stuff).~

Update: [No luck it seems](https://github.com/therealvio-org/ralphbot/pull/383). I am reverting back to my old config.

Update 2: This "update" I found was released last year. It's nearly the end of March, and I still think 2022 is this year 🤦‍♂️
